builtin/credential/oidc/path_config.go

```diff
@@ -0,0 +1,275 @@
+	if err != nil {
+		return logical.ErrorResponse(err.Error()), nil
+	}
+	if err := req.Storage.Put(jsonCfg); err != nil {
+		return logical.ErrorResponse(err.Error()), nil
+	}
+
+	return nil, nil
+}
+
+func (b *backend) pathConfigExistenceCheck(
+	req *logical.Request, d *framework.FieldData) (bool, error) {
+	cfg, err := b.Config(req.Storage)
+	if err != nil {
+		return false, err
+	}
+
+	return cfg != nil, nil
+}
+
+func (b *backend) oidcProviderForConfig(conf *ConfigEntry) (*oidc.Provider, error) {
+	// TODO(mwitkow): This creates a new Provider each time, making a request to the Discovery URL of the provider. Add caching per config "hash".
+	ctx := context.TODO()
+	if conf.IssuerCABundle != "" {
+		bundles, err := certPoolFromString(conf.IssuerCABundle)
+		if err != nil {
+			return nil, err
+		}
+		transport := &http.Transport{
+			TLSClientConfig: &tls.Config{
+				RootCAs: bundles,
+			},
+		}
+		httpClient := &http.Client{
+			Transport: transport,
+		}
+		ctx = oidc.ClientContext(ctx, httpClient)
+	}
+
+	return oidc.NewProvider(ctx, conf.IssuerUrl)
+}
+
+func (b *backend) validateAndExtractClaims(cfg *ConfigEntry, rawToken string) (userName string, groups []string, token *oidc.IDToken, err error) {
+	provider, err := b.oidcProviderForConfig(cfg)
+	if err != nil {
+		return "", []string{}, nil, err
+	}
+	if provider == nil {
+		return "", []string{}, nil, errors.New("OIDC provider not configured")
+	}
+	verifier := provider.Verifier(&oidc.Config{SkipClientIDCheck: true}) // note we verify the audience ourselves.
+	idToken, err := verifier.Verify(context.TODO(), rawToken)
+	if err != nil {
+		return "", []string{}, nil, errors.New("OIDC identity token error: " + err.Error())
+	}
+	if !containsAny(idToken.Audience, cfg.ClientIDs) {
+		b.Logger().Info("OIDC token has bad 'aud': %v", idToken.Audience)
+		return "", []string{}, nil, errors.New("OIDC identity token issued for unsupported client ID.")
+	}
+	claims := make(map[string]interface{})
+	if err := idToken.Claims(&claims); err != nil {
+		b.Logger().Warn("OIDC token claims parsing error: %v", err)
+		return "", []string{}, nil, errors.New("OIDC identity token claim parsing error.")
+	}
+	b.Logger().Debug("OIDC claims %v", claims) // TODO(mwitkow): Remove.
+
+	userNameClaim, ok := claims[cfg.UsernameClaim].(string)
+	if !ok {
+		b.Logger().Warn("OIDC token doesn't have username under expected claim. Claims: %v", claims)
+		return "", []string{}, nil, errors.New("OIDC identity token user claim parsing error.")
+	}
+	userName = userNameClaim
+	if cfg.GroupsClaim != "" {
+		groupNameClaim, ok := claims[cfg.GroupsClaim].([]interface{})
+		if !ok {
+			b.Logger().Warn("OIDC token doesn't have group name under expected claim.", "claimName", cfg.GroupsClaim, "allClaims", claims)
+			return "", []string{}, nil, errors.New("OIDC identity token group claim parsing error.")
+		}
+		for _, gInt := range groupNameClaim {
+			g, ok := gInt.(string)
+			if ok {
+				groups = append(groups, g)
```

Comment:
one thing we need to watch out for is groups conflicting with existing groups in vault. I don't know if plugin groups are prefixed but if not, it might be worth prefixing these strings with "oidc:" or something to namespaces them.